src/Parser/Core/Modules/StatTracker.js

```diff
@@ -0,0 +1,335 @@
+   */
+  get critRatingToPercent() {
+    return 1 / 40000;
+  }
+  get hasteRatingToPercent() {
+    return 1 / 37500;
+  }
+  get masteryRatingToPercent() {
+    switch (this.combatants.selected.spec) {
+      case SPECS.HOLY_PALADIN:
+        return 1 / 26667;
+      case SPECS.HOLY_PRIEST:
+        return 1 / 32000;
+      case SPECS.RESTORATION_SHAMAN:
+        return 1 / 13333;
+      case SPECS.ENHANCEMENT_SHAMAN:
+        return 1 / 13333;
+	    case SPECS.RESTORATION_DRUID:
+	      return 1 / 66667;
+      default:
+        throw new Error('Mastery hasn\'t been implemented for this spec yet.');
+    }
+  }
+  get versRatingToPercent() {
+    return 1 / 47500;
+  }
+  get avoidanceRatingToPercent() {
+    return 1 / 11000;
+  }
+  get leechRatingToPercent() {
+    return 1 / 23000;
+  }
+  get speedRatingToPercent() {
+    throw new Error('Speed hasn\'t been implemented yet.');
+  }
+
+  /*
+   * For percentage stats, the current stat percentage as tracked by this module.
+   */
+  get currentCritPercentage() {
+    return this.baseCritPercentage + (this.currentCritRating * this.critRatingToPercent);
+  }
+  get currentHastePercentage() {
+    return this.baseHastePercentage + (this.currentHasteRating * this.hasteRatingToPercent);
+  }
+  get currentMasteryPercentage() {
+    return this.baseMasteryPercentage + (this.currentMasteryRating * this.masteryRatingToPercent);
+  }
+  get currentVersPercentage() {
+    return this.baseVersPercentage + (this.currentVersRating * this.versRatingToPercent);
+  }
+  get currentAvoidancePercentage() {
+    return this.baseAvoidancePercentage + (this.currentAvoidanceRating * this.avoidanceRatingToPercent);
+  }
+  get currentLeechPercentage() {
+    return this.baseLeechPercentage + (this.currentLeechRating * this.leechRatingToPercent);
+  }
+  get currentSpeedPercentage() {
+    return this.baseSpeedPercentage + (this.currentSpeedRating * this.speedRatingToPercent);
+  }
+
+
+  on_toPlayer_changebuffstack(event) {
+    this._changeBuffStack(event);
+  }
+
+  on_toPlayer_changedebuffstack(event) {
+    this._changeBuffStack(event);
+  }
+
+  _changeBuffStack(event) {
+    const spellId = event.ability.guid;
+    const statBuff = this.constructor.STAT_BUFFS[spellId];
+    if(statBuff) {
+      // ignore prepull buff application, as they're already accounted for in combatantinfo
+      // we have to check the stacks count because Entities incorrectly copies the prepull property onto changes and removal following the application
+      if(event.oldStacks === 0 && event.prepull) {
+        debug && console.log(`StatTracker prepull application IGNORED for ${SPELLS[spellId] ? SPELLS[spellId].name : spellId}`);
+        return;
+      }
+      debug && console.log(`StatTracker: (${event.oldStacks} -> ${event.newStacks}) ${SPELLS[spellId] ? SPELLS[spellId].name : spellId}`);
+      this._changeStats(this._stats, statBuff, event.newStacks - event.oldStacks);
+    }
+  }
+
+  _changeStats(orig, change, factor) {
```

Comment:
seeing as this isn't reused *ever*, I think it would be cleaner to not have `orig` as a parameter but just directly set the value. Mutating parameters is generally undesirable for being hard to track and fragile, and if there's no reason at all to do it then I think we should avoid it.

Reply:
A holdover from an earlier design, I'll change it